src/App.tsx

```diff
@@ -1,24 +1,162 @@
+    setHasError(false);
+  }, []);
+
+  const setError = (errorTxt: string) => {
+    setErrorText(errorTxt);
+    setHasError(true);
+    setTimeout(resetError, 3000);
+  };
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(setTodos)
+      .catch((error) => {
+        setError(error.message);
+      });
+  }, []);
+
+  const onFilterChange = useCallback((filter: FilterBy) => {
+    setFilterBy(filter);
+  }, []);
+
+  const onAddTodo = async (title: string) => {
+    if (!/\S/g.test(title)) {
+      setError('Title can\'t be empty');
+
+      return false;
+    }
+
+    const newTodo = {
+      title,
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo({ ...newTodo, id: 0 });
+
+    try {
+      const addedTodo = await addTodo(newTodo);
+
+      setTodos(prevTodos => [...prevTodos, addedTodo]);
+    } catch (error) {
+      setError('Unable to add a todo');
+    } finally {
+      setTempTodo(null);
+    }
+
+    return true;
+  };
+
+  const onDeleteTodo = useCallback(async (todoId: number) => {
+    setLoadingTodoIds(prevIds => [...prevIds, todoId]);
+
+    try {
+      await deleteTodo(todoId);
+
+      setTodos(prevTodos => prevTodos.filter(todo => todo.id !== todoId));
+    } catch (error) {
+      setError('Unable to delete a todo');
+    } finally {
+      setLoadingTodoIds(prevIds => prevIds.filter(id => id !== todoId));
+    }
+  }, []);
+
+  const onUpdateTodo = useCallback(async (todoId: number, data: TodoPatch) => {
+    setLoadingTodoIds(prevIds => [...prevIds, todoId]);
+
+    try {
+      await updateTodo(todoId, data);
+
+      setTodos(prevTodos => prevTodos.map(todo => {
+        if (todo.id === todoId) {
+          return {
+            ...todo,
+            ...data,
+          };
+        }
+
+        return todo;
+      }));
+    } catch (error) {
+      setError('Unable to update a todo');
+    } finally {
+      setLoadingTodoIds(prevIds => prevIds.filter(id => id !== todoId));
+    }
+  }, []);
+
+  const filterTodos = (todosArray: Todo[], filter: FilterBy) => {
+    return todosArray.filter(({ completed }) => (
+      filter === FilterBy.Active ? !completed : completed
+    ));
+  };
+
+  let visibleTodos = todos;
+
+  if (filterBy !== FilterBy.All) {
+    visibleTodos = filterTodos(todos, filterBy);
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">React Todo App - Add and Delete</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <TodoForm
+          todos={todos}
+          onAddTodo={onAddTodo}
+          onUpdateTodo={onUpdateTodo}
+        />
+
+        {Boolean(todos.length) && (
```

Comment:
It's unnecessary to use Boolean in Boolean(todos.length). You can use todos.length directly to check if an array is empty or not.

Reply:
yes, you are right, then let's use `!!todos.length` or `todos.length > 0`